### PR TITLE
Add TypeScript declaration

### PR DESCRIPTION
**Pull request by jokester**
This `index.d.ts` should be sufficient for most library consumers in TypeScript.

There are missing points though. To name a few:
- less common but legal option names, like `escape-Script`
- `optGet / optSet` with numerical option id. They are hard to remember and unstable across versions so I didn't generate overloads for them.

### CODE REVIEW

File: README.md
Comment (gagern) [inline README.md]: Are you willing to maintain the TypeScript support in the future? If not, I guess I'd find a formulation which expresses that this is not an integral part of the codebase, particularly since I'm not sure I can keep it up.

Either way I guess I'd get rid of the `<3` eventually, as I prefer a drier style for public documentation and reserve hearts for personal communication. Unless you insist…

File: package.json
Comment (gagern) [inline package.json:6]: Feel free to also add yourself to the contributors section.

File: src/index.d.ts
Comment (gagern) [inline src/index.d.ts]: If it's not for public use, then better leave this out. I had included it in the API docs for two reasons: to explain that TidyOption *has* a constructor, which can be used for instanceof checks and so on, and to explicitely state that the constructor should not be used. Not including it in the TS library would strengthen that message I think.

File: src/index.d.ts
Comment (gagern) [inline src/index.d.ts]: The `getErrorLog` method is currently not documented, because it should not be needed for normal operations. Every error should either be returned from a sync call or passed to the callback of an async call. It's just for the sake of a unit test that I allow access to this, to ensure we're not generating additional errors that we fail to report. The method may disappear at any moment. Perhaps I should rename it to `_getErrorLog` to communicate that idea.

File: src/index.d.ts
Comment (gagern) [inline src/index.d.ts]: Better include this script with the codebase. Feel free to create a directory called `util` and put it there. Bonus points if you can rig up recreation of this portion as part of the prepublish phase.

File: src/index.d.ts
Comment (gagern) [inline src/index.d.ts]: The `optGet` part is incorrect: this version does return a string from the picklist. The `optSet` part is not exhaustive, as all of these integer-typed pick-value options accept integers or strings. Might be worth modeling.

File: src/index.d.ts
Comment (gagern) [inline src/index.d.ts]: If you specify the dict like this, people may conceivably pass conflicting options to the constructor. I'd rather use a white lie and omit all the camel-cased names. That helps keep the sources consistent and avoids conflicts.

File: test/test-decl.ts
Comment (gagern) [inline test/test-decl.ts]: Feel free to add these as dev dependencies, and to include these tests in the `npm test` machinery.

**Review by gagern — CHANGES_REQUESTED** [review]
Thenk you for contributing this proposal. There are a number of items we should discuss before merging this, but I like where this is going and am sure we can work out the details.

**Comment by gagern** [comment]
Oh, and one more thing: can you please conform to the two spaces of indentation I'm using?

File: README.md
Comment (jokester) [reply README.md]: I'm fine to maintain the TS support part (likely to use this library in a few more projects).

Feel free to change/remove the readme if you find it inconsistent :)

**Comment by jokester** [comment]
Still working 👷 , will let you know when I complete this list. I guess it will be before weekend.

- [x] make readme drier
- [x] add me to contributor array
- [x] prevent call to `TidyOption` (it can still be used in `instanceof`)
- [x] include `.d.ts` generate script and its deps
- [x] add a test case for type check
- [x] fix OptionDict: remove camel-cased names
- [x] indent with 2 spaces
- [x] remove `getErrorLog()` (undocumented & non-useful)
- [x] fix optGet: it may return a string from the picklist
- [x] fix optSet: integer-typed pick-value options also accept integers
- [x] examine my understanding for options
- [x] add more jsdoc to optGet

**Comment by jokester** [comment]
@gagern Pushed a few more commits. Mind checking them?

The generation script is included but not added to prepublish. I guess it safer to run `update-tsd` and have a human check it. Feel free to contact me when needed.

File: package.json
Comment (gagern) [inline package.json]: I wonder whether we should try to match the versions of these scoped packages here to those of the corresponding unscoped packages below. Any reasons in favor or against doing so?

File: util/tsconfig.json
Comment (gagern) [inline util/tsconfig.json]: Do we really need all of these settings? I'd prefer going with defaults as much as possible unless there is some major downside to doing so, in order to keep the number of settings that may need to be maintained low.

**Comment by jokester** [comment]
@gagern Haven't hear from you recently, but .. do you think there are something we should do before merging this?

File: README.md
Comment (gagern) [inline README.md]: Or easier: “… updated by running `npm run update-tsd`”.

File: src/index.d.ts
Comment (gagern) [inline src/index.d.ts]: Does this level of detail for the getters and setters really make sense? It does sound reasonable if you really are configuring these things manually, using string literals as keys. But those applications are easier to write using the options dictionary notation, I think. What happens if the configuration comes from a JSON file or similar? Is there some easy solution to make that work, too, or would that mean you'd have to type-cast the key from the JSON file to the enum of all the alternatives you have configured here? Would it make sense to add a catch-all definition up front? For the getter, that would be easy: `optGet(key: string): number | string | boolean` I think. For the setter that would become `optSet(key: string, value: number | string | boolean)` which would defy all the special cases, as this one case would capture anything not captured by the special cases, thus breaking the type checking there. So should we only do the specific getters, and the generic getter and setter? I'd like to see a generic get and set (using separately typed variables for key and value) demonstrated in the unit test.

File: util/gen-typescript-decl.ts
Comment (gagern) [inline util/gen-typescript-decl.ts]: Can you add a command line flag for “verify mode”, where you compare the content and exit with an error message and a non-zero exit status if there is a mismatch? Then we can have Travis run that mode to ensure the generated file stays in sync with the commited changes to the core library. This becomes moot if we autogenerate the code upon presubmit and leave it out of the repository.

File: src/index.d.ts
Comment (gagern) [inline src/index.d.ts]: Can you move this generated code to a separate file? I still believe not having the generated code in the repository, but generating it upon publication, would be the cleaner solution. That way we would not have to worry about inconsistent state, as there would be only a single source of truth, namely the tidy library. And with two files, it's easy to have one in the repository while ignoring the other. Automatically patching files is far more tricky in this respect.

**Review by gagern — COMMENTED** [review]
Sorry for the delay. Still some questions left.

File: package.json
Comment (jokester) [reply package.json]: This one is a bit tricky, `@types/` packages does not cover some of corresponding JS packages. Will see if I can fix or remove some packages.

File: util/tsconfig.json
Comment (jokester) [reply util/tsconfig.json]: It seems we can remove that file and still run the `.ts` code. Guess I should remove it.

File: util/gen-typescript-decl.ts
Comment (jokester) [reply util/gen-typescript-decl.ts]: I'm getting a little confused here: if we output the generated code to a separated file, and do not include it in the repo, is a "verify mode" still needed?

File: src/index.d.ts
Comment (jokester) [reply src/index.d.ts]: A value from JSON should still work without much effort: the TypeScript compiler (as of now) considers such a value to have `any` type, and excludes it from type check.

This is more of a limitation, those who read configure from JSON file cannot benefit from type checker.

Comment (jokester) [reply src/index.d.ts]: If I may say, adding generic getter / setter is not my favorite idea, mostly because allowing potentially incorrect option key is somehow against type checking. Can I ask what are some special cases that requires generic getter / setter?

File: package.json
Comment (gagern) [reply package.json]: I was mostly referring to those which have unscoped counterparts. No need to change the set of files, just the versions.

File: util/tsconfig.json
Comment (gagern) [reply util/tsconfig.json]: :+1: to that.

File: src/index.d.ts
Comment (gagern) [reply src/index.d.ts]: Correct me if I'm wrong, but here is what I imagine: When you read from JSON, the *value* would be of type `any`, but the *key* would be of type `string`, with no restriction as to what kind of string it would be. So you can't use the key from the JSON to configure these, unless you cast the `string` to something more generic. Right? I originally assumed you'd have to cast to the full list of enum values, but probably casting to `any` would be the more practical solution.

Another scenario might be query string parameters, which map string to string. As libtidy can parse strings, that's perfectly fine in terms of types, but would be problematic in a type-checked TypeScript.

File: util/gen-typescript-decl.ts
Comment (gagern) [reply util/gen-typescript-decl.ts]: True. That's what I meant with by “this becomes moot”.

File: package.json
Comment (jokester) [reply package.json]: Sorry to say this, but the `@types/mocha` package does not have a `3.x` version. The version `2.2` still it covers the mocha API we need.
Should I use this mismatched version, or remove this package and declare a minimal mocha API in place?

Comment (gagern) [reply package.json]: In that case better use the mismatched version. As far as I can tell, the TypeScript dependencies don't pull in the actual libraries, so we shouldn't see much impact from the mismatch unless we start using API which is different between those versions. Right?

A Pull Request against the DefinitelyTyped might be a good idea if you want to, but I think we don't block on that here.

File: src/index.d.ts
Comment (jokester) [reply src/index.d.ts]: Guess I got your idea, but this one didn't go quite well. It seems that when `optGet` has a generic overload, its return type will always be `number | string | boolean` regardless of the callsite, for example:

```ts
// Type 'string | number | boolean' is not assignable to type 'boolean'.
const markup: boolean = doc.optGet("markup");
```
(see [this commit](https://github.com/jokester/node-libtidy/commit/3dc9c75d88fa66636e7afc9f7bf0732d8fc2c928) for detail)

Do you think we can go with specialized getters and generic setters?

File: util/gen-typescript-decl.ts
Comment (jokester) [reply util/gen-typescript-decl.ts]: Understood. Guess I should add the generation script to `pretest` hook as well?

Comment (gagern) [reply util/gen-typescript-decl.ts]: Sounds good, yes.

**Comment by gagern** [comment]
Looks like a chicken and egg problem: you can't run the generator to build the options unless you already have the options build by the generator. One solution would be having the generator written in plain JavaScript. Another would be handling the option verification somewhere else, e.g. in a unit test.

**Comment by jokester** [comment]
Thanks for pointing out. I was able to reproduce the chicken-egg problem in VS Code editor (strangely, not when running the script). Guess changing the script to JavaScript is the safer fix.

File: src/index.d.ts
Comment (gagern) [reply src/index.d.ts]: Sorry, lost track of this PR. Is there any reason why you'd write the above in that way, instead of

```js
const markup: boolean = doc.options.markup;
```

That way you could use the dict for specific access, and use the getters for generic access, i.e. things like this:

```js
for (let o: doc.getOptionList())
  console.log({[o.name]: doc.optGet(o.name)})
```

Likewise, having all the various forms for values in the dictionary is harmful since you can't write things like

```js
const check: string = doc.opts.accessibility_check;
```

which I consider bad. Anyway, I finally got round to merging your PR, and did clean up the interface according to my taste in 0dc12f4b2b61d0543fe92d4abc4ab422534ee4ea.

Comment (jokester) [reply src/index.d.ts]: gagern: guess I did not consider that case. It looks much better with your fix, thanks!

**Pull request merged**
